### DISCUSSION TITLE
2D Voronoi cells

### DISCUSSION THREAD

**Discussion by CsatiZoltan** [MoiseRousseau/SALOME-Voronoi#3] [Ideas]
Hi,

What parameters should I pass to `vorpalite` to allow the computation of a 2D constrained Voronoi diagram? I.e. I would like the same functionality this plugin provides, but in 2D.

Thank you.

**Comment by MoiseRousseau** [comment]
Hi!

Unfortunately, the Vorpalite program does not permit to create 2D Voronoi diagram. You can see this line 401 of `main.cpp` located in your geogram folder in `src/bin/vorpalite`:

`CVT.set_volumetric(true);`

In any case, Vorpalite is set to volumetric mode. You can probably change this line with a `false` and recompile Geogram. However, I did not test it myself, and some modifications of the plugin will be necessary.

If case you don't want to modify the Vorpalite code, you can extrude your 2D mesh in 3D with a thin extent in Salome, then create a 3D Voronoi diagram and finally get its intersection with your 2D surface.

Moise

**Reply by CsatiZoltan** [reply]
Thanks for the tips. I tried the easiest idea, the one you proposed in the last paragraph.
I created a square, extruded it, but also kept the original square as a child in the extruded object, so that I can use that later during the meshing. Then I constructed the Voronoi cells with your plugin. After that, I tried to obtain the part that lies on the original square. It didn't work, whether I created a "Group on geometry" or a "Group on filter": it resulted in an empty group. This seems to make sense as the Voronoi mesh only contains volumes, no faces that could be selected. So the only way I see is indeed performing an intersection of the 3D Voronoi cells with the original square. However, I couldn't find that feature in Salome. Do you have an idea how I could do that?

**Reply by MoiseRousseau** [reply]
You have right, the Voronoi mesh created by the plugin only contains volumes. It is also no more connected to the SHAPER/GEOM module so you can't use the "Create Group on Geometry" option. If you want to create a group of faces, you first have to generate the boundary faces under "Modification / Create boundary elements" and then use a "Group on filter" with option "Belong to Geom".
So, for your 2D Voronoi mesh you have to:
1. Create the 2D mesh
2. Extrude it to make it 3D
3. Call the Voronoi mesher
4. Create the boundary faces with "Modification / Create boundary elements" 
5. Create a "Group on filter" with "Below to Geom" and select your original 2D geometry
Let me know if it's work.

**Reply by CsatiZoltan** [reply]
It works perfectly! Thanks for the steps. As this is a non-intrusive method (no compilation is needed by the user), it could easily be implemented. If you are interested, I can open a pull request once it is ready.

**Reply by MoiseRousseau** [reply]
If you want to contribute I will be happy to add your code to the plugin. I look forward your pull request!
